// @nibiruchain/nibijs/dist/msg/perp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerpMsgFactory = exports.isMsgPartialCloseEncodeObject = exports.isMsgDonateToEcosystemFundEncodeObject = exports.isMsgClosePositionEncodeObject = exports.isMsgOpenPositionEncodeObject = exports.isMsgMultiLiquidateEncodeObject = exports.isMsgRemoveMarginEncodeObject = exports.isMsgAddMarginEncodeObject = exports.perpTypes = exports.PERP_MSG_TYPE_URLS = void 0;
const tx_1 = require("@nibiruchain/protojs/dist/nibiru/perp/v2/tx");
const state_1 = require("@nibiruchain/protojs/dist/nibiru/perp/v2/state");
const chain_1 = require("../chain");
exports.PERP_MSG_TYPE_URLS = {
    MsgAddMargin: `/${tx_1.protobufPackage}.MsgAddMargin`,
    MsgRemoveMargin: `/${tx_1.protobufPackage}.MsgRemoveMargin`,
    MsgMultiLiquidate: `/${tx_1.protobufPackage}.MsgMultiLiquidate`,
    MsgMarketOrder: `/${tx_1.protobufPackage}.MsgMarketOrder`,
    MsgClosePosition: `/${tx_1.protobufPackage}.MsgClosePosition`,
    MsgDonateToEcosystemFund: `/${tx_1.protobufPackage}.MsgDonateToEcosystemFund`,
    MsgPartialClose: `/${tx_1.protobufPackage}.MsgPartialClose`,
};
exports.perpTypes = [
    [exports.PERP_MSG_TYPE_URLS.MsgAddMargin, tx_1.MsgAddMargin],
    [exports.PERP_MSG_TYPE_URLS.MsgRemoveMargin, tx_1.MsgRemoveMargin],
    [exports.PERP_MSG_TYPE_URLS.MsgMultiLiquidate, tx_1.MsgMultiLiquidate],
    [exports.PERP_MSG_TYPE_URLS.MsgMarketOrder, tx_1.MsgMarketOrder],
    [exports.PERP_MSG_TYPE_URLS.MsgClosePosition, tx_1.MsgClosePosition],
    [exports.PERP_MSG_TYPE_URLS.MsgDonateToEcosystemFund, tx_1.MsgDonateToEcosystemFund],
    [exports.PERP_MSG_TYPE_URLS.MsgPartialClose, tx_1.MsgPartialClose],
];
const isMsgAddMarginEncodeObject = (encodeObject) => encodeObject.typeUrl ===
    exports.PERP_MSG_TYPE_URLS.MsgAddMargin;
exports.isMsgAddMarginEncodeObject = isMsgAddMarginEncodeObject;
const isMsgRemoveMarginEncodeObject = (encodeObject) => encodeObject.typeUrl ===
    exports.PERP_MSG_TYPE_URLS.MsgRemoveMargin;
exports.isMsgRemoveMarginEncodeObject = isMsgRemoveMarginEncodeObject;
const isMsgMultiLiquidateEncodeObject = (encodeObject) => encodeObject.typeUrl ===
    exports.PERP_MSG_TYPE_URLS.MsgMultiLiquidate;
exports.isMsgMultiLiquidateEncodeObject = isMsgMultiLiquidateEncodeObject;
const isMsgOpenPositionEncodeObject = (encodeObject) => encodeObject.typeUrl ===
    exports.PERP_MSG_TYPE_URLS.MsgMarketOrder;
exports.isMsgOpenPositionEncodeObject = isMsgOpenPositionEncodeObject;
const isMsgClosePositionEncodeObject = (encodeObject) => encodeObject.typeUrl ===
    exports.PERP_MSG_TYPE_URLS.MsgClosePosition;
exports.isMsgClosePositionEncodeObject = isMsgClosePositionEncodeObject;
const isMsgDonateToEcosystemFundEncodeObject = (encodeObject) => encodeObject.typeUrl ===
    exports.PERP_MSG_TYPE_URLS.MsgDonateToEcosystemFund;
exports.isMsgDonateToEcosystemFundEncodeObject = isMsgDonateToEcosystemFundEncodeObject;
const isMsgPartialCloseEncodeObject = (encodeObject) => encodeObject.typeUrl ===
    exports.PERP_MSG_TYPE_URLS.MsgPartialClose;
exports.isMsgPartialCloseEncodeObject = isMsgPartialCloseEncodeObject;
// ----------------------------------------------------------------------------
class PerpMsgFactory {
    static removeMargin(msg) {
        return {
            typeUrl: exports.PERP_MSG_TYPE_URLS.MsgRemoveMargin,
            value: tx_1.MsgRemoveMargin.fromPartial(msg),
        };
    }
    /**
     * Returns a 'TxMessage' for adding margin to a position
     *
     * @static
     * @param {MsgAddMargin} msg - Message to add margin
     * @returns {TxMessage} - formatted version of MsgAddMargin
     */
    static addMargin(msg) {
        return {
            typeUrl: exports.PERP_MSG_TYPE_URLS.MsgAddMargin,
            value: tx_1.MsgAddMargin.fromPartial(msg),
        };
    }
    static liquidate(msg) {
        return {
            typeUrl: exports.PERP_MSG_TYPE_URLS.MsgMultiLiquidate,
            value: tx_1.MsgMultiLiquidate.fromPartial(msg),
        };
    }
    static openPosition(msg) {
        const { quoteAssetAmount, baseAssetAmountLimit, leverage } = msg;
        const pbMsg = {
            sender: msg.sender,
            pair: msg.pair,
            quoteAssetAmount: (0, chain_1.toSdkInt)(quoteAssetAmount),
            baseAssetAmountLimit: (0, chain_1.toSdkInt)(baseAssetAmountLimit !== null && baseAssetAmountLimit !== void 0 ? baseAssetAmountLimit : 0),
            leverage: (0, chain_1.toSdkDec)(leverage.toString()),
            side: msg.goLong ? state_1.Direction.LONG : state_1.Direction.SHORT,
        };
        return {
            typeUrl: exports.PERP_MSG_TYPE_URLS.MsgMarketOrder,
            value: tx_1.MsgMarketOrder.fromPartial(pbMsg),
        };
    }
    static closePosition(msg) {
        return {
            typeUrl: exports.PERP_MSG_TYPE_URLS.MsgClosePosition,
            value: tx_1.MsgClosePosition.fromPartial(msg),
        };
    }
    static partialClosePosition(msg) {
        return {
            typeUrl: exports.PERP_MSG_TYPE_URLS.MsgPartialClose,
            value: tx_1.MsgPartialClose.fromPartial(msg),
        };
    }
    static donateToPerpEF(msg) {
        return {
            typeUrl: exports.PERP_MSG_TYPE_URLS.MsgDonateToEcosystemFund,
            value: tx_1.MsgDonateToEcosystemFund.fromPartial(msg),
        };
    }
}
exports.PerpMsgFactory = PerpMsgFactory;
