// @nibiruchain/nibijs/dist/query/inflation.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupInflationExtension = void 0;
const stargate_1 = require("@cosmjs/stargate");
const query_1 = require("@nibiruchain/protojs/dist/nibiru/inflation/v1/query");
const setupInflationExtension = (base) => {
    const rpcClient = (0, stargate_1.createProtobufRpcClient)(base);
    const queryService = new query_1.QueryClientImpl(rpcClient);
    return {
        inflation: {
            period: () => __awaiter(void 0, void 0, void 0, function* () {
                const req = query_1.QueryPeriodRequest.fromPartial({});
                const resp = yield queryService.Period(req);
                return resp;
            }),
            epochMintProvision: () => __awaiter(void 0, void 0, void 0, function* () {
                const req = query_1.QueryEpochMintProvisionRequest.fromPartial({});
                const resp = yield queryService.EpochMintProvision(req);
                return resp;
            }),
            skippedEpochs: () => __awaiter(void 0, void 0, void 0, function* () {
                const req = query_1.QuerySkippedEpochsRequest.fromPartial({});
                const resp = yield queryService.SkippedEpochs(req);
                return resp;
            }),
            circulatingSupply: () => __awaiter(void 0, void 0, void 0, function* () {
                const req = query_1.QueryCirculatingSupplyRequest.fromPartial({});
                const resp = yield queryService.CirculatingSupply(req);
                return resp;
            }),
            inflationRate: () => __awaiter(void 0, void 0, void 0, function* () {
                const req = query_1.QueryInflationRateRequest.fromPartial({});
                const resp = yield queryService.InflationRate(req);
                return resp;
            }),
            params: () => __awaiter(void 0, void 0, void 0, function* () {
                const req = query_1.QueryParamsRequest.fromPartial({});
                const resp = yield queryService.Params(req);
                return resp;
            }),
        },
    };
};
exports.setupInflationExtension = setupInflationExtension;
